KAFKA-3816: Add MDC logging to Connect runtime

rhauch opened the pull request:
See https://cwiki.apache.org/confluence/display/KAFKA/KIP-449%3A+Add+connector+contexts+to+Connect+worker+logs

Added LoggingContext as a simple mechanism to set and unset Mapped Diagnostic Contexts (MDC) in the loggers to provide for each thread useful parameters that can be used within the logging configuration. MDC avoids having to modify lots of log statements, since the parameters are available to all log statements issued by the thread, no matter what class makes those calls.

The design intentionally minimizes the number of changes to any existing classes, and does not use Java 8 features so it can be easily backported if desired, although per this KIP it will be applied initially only in AK 2.3 and later and must be enabled via the Log4J configuration.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-3816: Add MDC logging to Connect runtime.

Added LoggingContext as a simple mechanism to set and unset Mapped Diagnostic Contexts (MDC) in the loggers to provide for each thread useful parameters that can be used within the logging configuration. MDC avoids having to modify lots of log statements, since the parameters are available to all log statements issued by the thread, no matter what class makes those calls.

The design intentionally minimizes the number of changes to any existing classes, and does not use Java 8 features so it can be easily backported if desired.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,229 @@
+        /**
+         * The name of the Mapped Diagnostic Context (MDC) key that defines the type of <i>connector</i>.
+         */
+        public static final String CONNECTOR_TYPE = "connector.type";
```
it's not user-friendly, but most MDC Contexts tend to avoid verbose keys (wherever possible values) since it can easily fill up the logs.

rhauch replied:
This is the name of the MDC property, not what is included in the log files.

rhauch replied:
Changed with the latest commit.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,229 @@
+        /**
+         * The name of the Mapped Diagnostic Context (MDC) key that defines the type of <i>connector</i>.
+         */
+        public static final String CONNECTOR_TYPE = "connector.type";
```
From the name it appears to be source vs sink but it doesn't look to be so.

rhauch replied:
Yeah, that should probably be changed. It is not "source" vs "sink", but rather a shortened name of the connector class. I'm actually not convinced this adds much value in the log compared to the connector name, and I've thought about removing it altogether to simplify things. 

The only reason I've kept it is because one could then easily filter the log to get all logs associated with a type of connector. Is that really useful?

rhauch replied:
Changed with the latest commit.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,229 @@
+
+    private static final String NULL_VALUE = null;
+
+    final Map<String, String> previous = new HashMap<>();
```
why do we have to maintain this state when MDC. getCopyOfContextMap can get us the context. In fact, I think we should use that when we create a new MDC Context for the Worker Tasks.

rhauch replied:
Because when you use `getCopyOfContextMap` you get all of the properties and you still have to unset all of them and then put all of what's in the copy. However, that doesn't work as well if you're incrementally adding/removing a subset of the contexts, and it doesn't work if there are other components (e.g., connectors) that are adding contexts after we have.

The current approach is a bit more complicated, but it should be a lot easier to be completely independent of all other modifications of the MDC context map.

mageshn reviewed (commented):
Thanks, @rhauch. I think this will be a great and long awaited addition to Connect.  I took a high-level pass and had some questions.

rhauch commented:
@ewencp, @kkonstantine: any feedback?

rhauch committed:
KAFKA-3816: Minor changes to the connector kind MDC parameter

Renamed the `connector.type` MDC parameter to `connector.class`, since the value represents not “source” or “sink” (which we refer to elsewhere as the _connector type_) but rather a shortened version of the class name.

rhauch commented:
@mageshn, I rebased to the latest from `trunk` and changed the `connector.type` parameter name to `connector.class`, since as you mention we use "_connector type_" elsewhere to refer to whether the connector is a "source" or a "sink".

I also updated the `connect-lo4j.properties` file to include a commented-out format statement that uses these. Note that users will still have to explicitly enable this in their logging configuration in order to even see these in their log statements, so this change is backward compatible.

rhauch committed:
KAFKA-3816: Simplified the LoggingContext functionality

Redesigned the new LoggingContext functionality to remove all local state and to simplify the usage and implementation. Added a `LoggingContext.clear()` method that will remove all Connect-specific MDC parameters from the current thread’s MDC context, and used this in several places to ensure that the parameters don’t leak to other uses of the same thread.

Also removed the connector class MDC parameter, since this doesn’t add much value compared with the connector name.

rhauch commented:
@mageshn, I took your advice and simplified the `LoggingContext` class to remove any instance-based functionality. As before, the static methods are used in several key places within Connect, but no longer return a `LoggingContext` instance that can be used to clean up the context. Instead, a new `clean()` static method is introduced to allow callers to remove the Connect-specific MDC parameters. Note that Connect uses a pool of threads, and cleaning the MDC parameters after leaving a context is still warranted to ensure those parameters do not leak should that thread be used in a different context.

I'd appreciate another review.

mageshn reviewed (commented):
Thanks, @rhauch. Overall, this LGTM. Should we possibly add some documentation about these MDC keys that are being used by the framework? This could potentially help connectors to not override these. I can't think of a way to prevent from something like this completely.

rhauch commented:
@mageshn I'm not sure where we would document those MDC context parameters. The https://kafka.apache.org/ site doesn't appear to document any of the logging configurations. I did add them as comments in the [connect-log4j.properties](https://github.com/apache/kafka/pull/5743/files#diff-8046fa21815eed40617fc7d85c292033R28) file, which given the lack of logging documentation seems sufficient IMO.

mageshn reviewed (approved):
That sounds good to me @rhauch. I think this would be a super useful feature. LGTM

rhauch commented:
@ewencp or @hachikuji, would you mind reviewing and merging if acceptable?

rhauch commented:
@ewencp or @hachikuji, reminder to review, please.

rhauch committed:
KAFKA-3816: Removed connector type from test helper methods

rhauch commented:
The JDK 8 / Scala 2.11 failure is unrelated:

    kafka.api.AdminClientIntegrationTest.testMinimumRequestTimeouts

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java:
```diff
@@ -79,7 +80,12 @@ public void schedule(final ConnectorTaskId id, final WorkerSourceTask workerTask
             public void run() {
-                commit(workerTask);
+                try {
+                    LoggingContext.forOffsets(id);
```
I see from previous comments that there used to be `LoggingContext` instances. Given current structure of code, I can see how getting rid of instances and uses `static` methods makes sense, but should we consider still using instances and changing `clear` to `close()` so we can just use try-with-resources? That seems less error-prone than this setup where it'd be easy to forget to put in a `clear` call.

rhauch replied:
Agreed. I think my previous simplification went too far, and AutoCloseable does indeed make it much less error prone. Another benefit of using a try-with-resources and AutoCloseable is that the instance can capture and restore the complete MDC parameter state, and this makes nesting logging contexts that much easier and less error-prone.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -220,6 +221,7 @@ public boolean startConnector(
         final WorkerConnector workerConnector;
         ClassLoader savedLoader = plugins.currentThreadLoader();
         try {
+            LoggingContext.forConnector(connName);
```
How did we decide what to change logging context for vs not? For example, `connectorTaskConfigs` later in this file doesn't do this, but it will make a call into the connector which could potentially do logging without providing context as to which *instance* is logging. Similarly, `stopConnector` doesn't seem to switch logging context. (in particular, I found those by checking for where we're doing plugin swapping, which seems a likely indicator we want to change contexts)

rhauch replied:
Good catch. I've tried to not change the herder, which is where some of the plugin swapping is done, but `connectorTaskConfigs` and `stopConnector` and `stopTask` all should have switched logging contexts.

I also adjusted where the logging context is created in some of the other methods, to be more inclusive of the whole Worker method, rather than reusing an existing try-block to minimize changes.

rhauch replied:
@ewencp, ah, I remember why I moved away from AutoCloseable: In a try-with-resources statement, any `catch` or `finally` block is run _after_ the resources declared have been closed.

I still think it's cleaner, but it may require an outer try-with-resources block outside of an existing try-catch-finally block to ensure that all log messages within the existing catch and finally blocks are logged in the same context.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,121 @@
+        WORKER("worker"),
+
+        /**
+         * The {@link Parameters#CONNECTOR_SCOPE scope} value for {@link org.apache.kafka.connect.connector.Task}
```
it looks like the plain `connector` context is implicit, any reason not to just make it explicit?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,121 @@
+    public static void forConnector(String connectorName) {
+        MDC.put(Parameters.CONNECTOR_NAME, connectorName);
+        MDC.remove(Parameters.CONNECTOR_TASK);
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.WORKER.value());
```
actually, i guess i see here that for the connector itself we'd use `WORKER`. why is that?

I know we swap out the context such that some logs would be coming from the worker (e.g. the log messages indicating we are going to instantiate the connector when we start it), but that's also true for tasks where we scope to `task`.

when reading through this initially and i saw worker, I assumed anything logged for that would indicate something that might not invoke the connector/task at all but is done on behalf of them (initially i was thinking offset commit for this but that would invoke; possibly something having to do with configs or statuses?), or where there's not a connector name yet (e.g. validations).

rhauch replied:
I guess because the WorkerTask is doing more than the connector is, but it is doing it for a particular named connector. So when viewing the log, I'd want to see all of the worker-related logic (whether or not it's Connect code or Connector code) in the "worker" context, and all task-related logic (whether it's Connect's WorkerTask or connector task) in the "task" context. The class names in each log message tell me exactly where that code comes from.

Based upon your previous question, is your main question why we use "worker" and not "connector" for the scope? I think I had used "connector" at some point, and think my primary reason was that "connector" scope was used for much more than just what the individual `Connector` instance was doing (even how a user might think of it). Plus "worker" was shorter, which was a small benefit.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @param id the connector task ID; may not be null
+     */
+    public static void forOffsets(ConnectorTaskId id) {
```
is this distinction going to be useful to users or us? i know it operates in a separate thread, but so do some other things (e.g. stopping a source connector expects multi-threaded).

rhauch replied:
At least for us I think it does add clarity because these are called by the offset commit thread, not the connector's worker thread. Sometimes I've wanted to just find the messages that were written by all of the code running on the offset commit thread for a particular source connector, and it's not easy to do.

I'm less convinced this will be really obvious to users who are less familiar with the behaviors of Connect. Users do often see log messages related to offset commits and they're not sure whether that affects their source connector and the producer. However, I don't think a separate scope for offsets is detrimental.

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,121 @@
+        Scope(String value) {
+            this.value = value;
+        }
+        public String value() {
```
since value() is associated with a java specific thing for enums, can we call this function something else? maybe `type()`?

rhauch replied:
I've switched to `toString()` which is better anyway since it can be overridden and it should be used when you want a string form of the enum (which we do) as opposed to the name.

wicknicks commented on connect/runtime/src/test/resources/log4j.properties:
```diff
@@ -18,7 +18,22 @@ log4j.rootLogger=OFF, stdout
 
 log4j.logger.org.reflections=ERROR
 log4j.logger.org.apache.kafka=ERROR
+log4j.logger.org.apache.kafka.connect.util.LoggingContextTest=INFO
```
were we planning to revert this before merging?

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,121 @@
+    public static void clear() {
+        MDC.remove(Parameters.CONNECTOR_NAME);
+        MDC.remove(Parameters.CONNECTOR_TASK);
+        MDC.remove(Parameters.CONNECTOR_SCOPE);
```
just wondering why we need this method? AFAICT, we don't reuse threads in Connect. If a thread is dedicated for a particular task (as a task or offset committer), it is dedicated for that task throughout its lifetime. if we could get away without needing this method, it would simplify the changes. WDYT?

rhauch replied:
You're right that it's not often necessary. Where we do not switch threads, the extra overhead of resetting the MDC parameters is negligible. But the offset commit thread does switch logging contexts, and IMO it is simply a good practice to encourage restoring resources. 

Per @ewencp's suggestion, I went back to AutoCloseable so that try-with-resources works and reduces the risk of coding errors.

wicknicks replied:
The heavy lifting in `SourceTaskOffsetCommitter` is delegated back to the WorkerSourceTask though, which doesn't help us with MDC but does bring up if we want to use LogContext instead of MDC for AK? Other components in AK (Kafka, Streams and Clients) seem to use [LogContext](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/LogContext.java#L37-L39) that adds a prefix on every SLF4J logger. This is already being used by some bit of Connect (for example., in WorkerCoordinator). This would avoid having to micromanage thread context, and also keeps the platform uniform.

hachikuji replied:
Yes, I had the same thought as @wicknicks. We preferred `LogContext` over MDC because our needs were basic and the API was simpler.

rhauch replied:
@wicknicks, @hachikuji: indeed `LogContext` is simpler, but:
1. its prefix is fixed; and
2. it only affects log messages written by `Logger` instances instantiated with the `LogContext`.

So for a Kafka client, this works great because the client has a single ID and the AK code that creates the `Logger` can be changed to use `LogContext`.

However, with Connect the connector implementations will not change how they create their `Logger` instances. And, we do have some code where a single thread does switch contexts from one connector/task to another. Given these, I don't see how `LogContext` will work. 

This PR uses MDC, and this will add the additional context (via MDC parameters) to _all_ log messages written by _all_ components -- including connector implementations. MDC works this way because it is thread-based, not based upon the `Logger` instance that is being used, so switching contexts within a single thread is easy to do.

wicknicks replied:
alright cool -- yeah LogContext won't work if we want to tag message from plugins.

wicknicks reviewed (commented):
thanks, @rhauch! this is super useful.

rhauch committed:
KAFKA-3816: Changed to use try-with-resources blocks for logging contexts

Make LoggingContext implement AutoCloseable so that they can be created as resources within try-with-resources blocks and automatically cleaned up at the end of the try-with-resources. However, because resources are cleaned up before the `catch` and `finally` blocks, it was often necessary to _add_ a new try-with-resources rather than reuse an existing try-catch-finally block so that log messages within the `catch` and `finally` blocks are still logged in the same context.

rhauch commented:
@ewencp, @wicknicks: thanks for the review. Answered most questions/comments inline above, and updated the proposed changes based upon your feedback, and would appreciate another review. (Rebased on trunk and forced pushed to the PR.)

rhauch commented:
@ewencp, @wicknicks: would you mind reviewing the latest changes, and approving if you have no other concerns?

rhauch commented on config/connect-log4j.properties:
```diff
@@ -18,6 +18,19 @@ log4j.rootLogger=INFO, stdout
+# Use %X{connector.scope} for the scope within a connector (e.g., "worker", "task", or "offsets")
+# Use %X{connector.task} for the 1-based number of the task, or blank if not within a connector task
+#
+#log4j.appender.stdout.layout.ConversionPattern=[%d] %p %X{connector.name}|%X{connector.scope}%X{connector.task} %m (%c:%L)%n
```
Should we enable this by default in the provided `connect-log4j.properties`?

ewencp replied:
log format is a public interface, if we do something like this we'd need to consider the compatibility implications. (we're already probably more lax in log message changes than we really should be...)

rhauch replied:
Ok, that's why I had only included it as a comment.

wicknicks replied:
In this example, should we add a `|` between scope and task?

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,142 @@
+        LoggingContext context = new LoggingContext();
+        MDC.put(Parameters.CONNECTOR_NAME, connectorName);
+        MDC.remove(Parameters.CONNECTOR_TASK);
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.WORKER.toString());
```
is there a reason to retain these different components? I think it might be easier if we only declared a single `%X{connect.location}`, and internally format it as we see fit. That way we can omit metadata when we think is appropriate.  Couple of other advantages I can think of:
- easier for users to reason about
- when we add the formatting bits (`|` in the examples here), these show up in the logs no matter what the content of the MDC map. For example: `[2019-02-19 16:49:47,580] (main) INFO || Scanning for plugin classes.`.
- optional variables will be more easy to manage (right now, tasks numbers are not present in some cases).
- In future, if we decide to add more variables, then those will show up in the logs without users having to muck around with their `log4j.properties` files.

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,142 @@
+ * <p>{@link LoggingContext} instances should be created in a try-with-resources block to ensure that the logging context is properly
+ * closed. The only exception is the logging context created upon thread creation that is to be used for the entire lifetime of the thread.
+ */
+public final class LoggingContext implements AutoCloseable {
```
should this class be called `ConnectLoggingContext` to avoid similarities it from `LogContext`.

C0urante replied:
@wicknicks can you clarify how a similarity in name here would be problematic? Unless the two are used in the same class it doesn't seem like that'd be an issue; you can always use the package name to help differentiate between the two when figuring out which one to import.
Plus, although it's unlikely, this class may come in handy in other contexts besides Connect.

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,142 @@
+    public void close() {
+        MDC.clear();
+        if (previous != null && !previous.isEmpty()) {
+            MDC.setContextMap(previous);
```
I think this would break if a thread directly sets `MDC` keys (maybe in a Connector) in between a `LoggingContext.for..()`  and its `close()` call. For example: 

```
log.info("blah 1");  // outputs "blah 1"
try (LoggingContext ignored = LoggingContext.forConnector("jello")) {
    MDC.put("hello", "world");
    log.info("blah 2"); // outputs "world jello|worker| blah 2"
}
log.info("blah-3"); // outputs *only* "blah-3", the "hello", "world" entry is removed from MDC map.
```

Maybe we want to just remove the keys we set to be safer overall?

rhauch replied:
That was considered in an earlier version, and @mageshn suggested we do this.

mageshn replied:
The only time a LoggingContext should be cleared is when a thread gets repurposed for a different reason. IIRC, the LoggingContext was not `Closeable` when I provided the original comment. If the general agreement is that it should be `Closeable` then we got find a way to manage the state then. IMO, I think it could become complicated trying to manage the state when the MDC is already doing in in threadLocal. One another interesting idea could be invoke close some how as part of the Plugin class loader swap, while not intuitive, it will ensure that we are clearing the context everytime its necessary. Thoughts?

wicknicks replied:
I think the source of the confusion arises from LoggingContext.close() calling MDC.clear. Removing specific keys instead would solve the problem.

rhauch replied:
Actually, the purpose of `LoggingContext` is to essentially capture the MDC state such that it can be restored at a later time when `LoggingContext.close()` is called. The purpose is not just to manage the now one MDC parameter in which we're storing the Connect context. The `close()` method does call `MDC.clear()`, but then it also sets the MDC state to the same parameters that were found when the `LoggingContext` constructor was called.

Consider when this is used to set up a context under which a connector implementation method is called, and that connector implementation sets an MDC parameter. With the current implementation, when the logging context is closed, any changes to the MDC state will have been cleaned up properly, and the MDC state restored to exactly what it was when the context was first created.

C0urante replied:
@wicknicks @rhauch @mageshn I think this behavior is acceptable. We may want to document this logging behavior somewhere so that nobody tries to set an `MDC` key in, e.g., `Connector.start()` and then refer back to it in `Connector.stop()`. However, at least from my POV it seems unreasonable and unintuitive to expect that to work in the first place. 

@rhauch I like the `shouldAllowNestedLoggingContexts` test you've written. As long as we agree that the behavior we're discussing here is acceptable, could you either add a few extra `MDC` keys inside one of the nested contexts and verify that they're cleared once the context is closed, or add a new test similar to `shouldAllowNestedLoggingContexts` that deals specifically with the wiping of extra `MDC` keys once a context is closed?

rhauch replied:
I think I see @wicknicks's point, though, and I think the current implementation might pose a problem. Consider an implementation of `SourceTask` that sets up an MDC parameter in the `start(...)` method and remove it in the `stop(...)` method, but wants any log messages written by this task class to include that MDC parameter. See [this real-world example](https://github.com/debezium/debezium/blob/9b405aa84d8391a7ccf883184d6e8c67bac8bed3/debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java).

Connect currently runs each task in a separate thread via the `Worker.executorService`, and all of the task methods are called from within that thread. So I think it's safe to not restore the MDC state to exactly what it was when the LoggingContext was stopped, as suggested by @wicknicks.

However, when the task ends, it's possible the executor service will reuse the thread, so we need to clean that in `WorkerTask.run()`, just in case the thread's MDC parameters contain state from a prior task.

wicknicks commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -277,35 +280,37 @@ public boolean isSinkConnector(String connName) {
-                if (connOriginals.containsKey(SinkTask.TOPICS_CONFIG)) {
-                    taskConfig.put(SinkTask.TOPICS_CONFIG, connOriginals.get(SinkTask.TOPICS_CONFIG));
+        try (LoggingContext loggingContext = LoggingContext.forConnector(connName)) {
+            log.trace("Reconfiguring connector tasks for {}", connName);
```
nit: when we put such large blocks within a try-finally block, do you think it would be useful to move the block to a separate private method, so they are more easy to read?

rhauch replied:
I guess in this case I don't believe that moving to a separate private method is that useful/beneficial.

wicknicks reviewed (commented):
took another pass at this. some comments.

rhauch committed:
KAFKA-3816: Changed to use a single MDC parameter

A single MDC parameter is used for Connect, with the value being generated when the LoggingContext is created from various inputs. This retains the same information, but exposes only a single parameter to the user. Note that this does not allow the user to control the format of the additional context strings, but that’s probably an advantage since we can standardize the format and the content.

rhauch commented:
Rebased, and updated to reflect the updated proposal also described in [KIP-449](https://cwiki.apache.org/confluence/display/KAFKA/KIP-449%3A+Add+connector+contexts+to+Connect+worker+logs).

rhauch committed:
KAFKA-3816: Added ‘-‘ to Connect logging context and enabled the context in Connect tests

rhauch commented:
Incorporated a few changes as suggested by @kkonstantine in the [KIP-449](https://cwiki.apache.org/confluence/display/KAFKA/KIP-449%3A+Add+connector+contexts+to+Connect+worker+logs) discussion.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/LoggingContextTest.java:
```diff
@@ -0,0 +1,181 @@
+    protected void assertMdc(String connectorName, Integer taskId, Scope scope) {
+        String context = MDC.get(LoggingContext.CONNECTOR_CONTEXT);
+        if (context != null) {
+            assertEquals(
```
This is a little hard to read. A vanilla `if` statement may be better:

```
if (connectorName != null) {
    assertTrue(context.startsWith("[" + connectorName));
}
```

rhauch replied:
But that's not equivalent. It'd need to be:
```
if (connectorName != null) {
    assertTrue(context.startsWith("[" + connectorName));
} else {
    assertFalse(context.startsWith("[" + connectorName));
}
```
IMO, that's too verbose.

C0urante replied:
@rhauch 
I don't believe that's the effect the code here has. In the method call:
```java
assertEquals(
  connectorName != null,
  connectorName != null ? context.startsWith("[" + connectorName) : false
);
```

if `connectorName` is null, then both arguments are guaranteed to evaluate to `false`.

I think your intent may have been something like this:
```java
assertEquals(
  connectorName != null,
  context.startsWith("[" + connectorName)
);
```

which would probably be acceptable, but may also benefit from a `message` that clarifies the expected behavior, possible something like `"Context should begin with connector name if and only if connector is non-null"`

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java:
```diff
@@ -0,0 +1,211 @@
+        sb.append(connectorName);
+        if (taskNumber != null) {
+            // There is a task number, so this is a task
+            sb.append("|");
```
Nit: may want to just move this line outside the two `if` statements since it's executed in both branches and IIUC exactly one of the branches is intended to run every time this method is invoked.

rhauch replied:
It's actually possible for the scope to be OFFSET (for example) but for the task number to be non-null (e.g., the offsets for connector "foo" task 3). In this case, both if-blocks will be entered, and moving the `sb.append("|");` out of the if blocks will not produce the correct output.

C0urante replied:
Agh, sorry, missed that. Thanks for the clarification.

C0urante reviewed (approved):
Thanks for the PR, @rhauch! I've left a few comments but nothing major, feel free to address at your discretion.

rhauch committed:
KAFKA-3816: No longer clear the MDC state when LoggingContext is closed

The previous approach had a potential bug: if a connector’s task were to set MDC parameters in its `start()` method (after a `LoggingContext` instance is created in the framework), then the `LoggingContext.close()` call within the framework will remove the connector’s MDC parameters before the framework creates a new `LoggingContext` and calls the task's `put(...)` or `poll(...)` method. The change is to only clear out the MDC parameters when the task starts, at which point there any MDC parameters set by a previous connector’s task are removed. Note that tasks are executed with an ExecutorService, which means its possible (depending upon the implementation) that some threads are reused.

rhauch reviewed (commented):
Thanks, @C0urante. Left a few responses to your comments. I've also updated the PR to address the potential bug that @wicknicks brought up, where an MDC parameter set by a connector's task in the `start()` method (after a `LoggingContext` instance is created in the framework) will be removed by the `LoggingContext.close()` before the task's `put(...)` or `poll(...)` method is called.

rhauch committed:
KAFKA-3816: Minor improvements to unit test, per review comments

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -214,38 +215,40 @@ public boolean startConnector(
-        workerMetricsGroup.recordConnectorStartupSuccess();
+            log.info("Finished creating connector {}", connName);
+            workerMetricsGroup.recordConnectorStartupSuccess();
+        }
```
The bulk of the changes in this class are indentation due to wrapping in a new try-catch.

guozhangwang reviewed (approved):
LGTM! (I like the way MDC is used).

@rhauch Could you rebase?

rhauch commented:
@guozhangwang thanks for the review! Rebased on latest.
